### PR TITLE
8320707: Virtual thread test updates

### PR DESCRIPTION
**Pull request by AlanBateman**
A lot of test changes have accumulated in the loom repo, this includes both new tests and updates to existing tests. Some of these updates can be brought to the main line. This update brings over:

- The existing tests for pinning use synchronized blocks. In preparation for changes to allow carrier thread be released when a virtual thread parks holding a monitor or blocks on monitorenter, these tests are changed to pin by having a native frame on the stack. This part includes test infrastructure to make it easy to add more tests that do operations while pinned. The tests still test what they were originally created to test of course.

- The test for the JFR jdk.VirtualThreadPinned event is refactored to allow for additional cases where the event may be reported.

- ThreadAPI is expanded to cover test for uncaught exception handling.

- GetStackTraceWhenRunnable is refactored to not use a Selector, otherwise this test will be invalidated when blocking selection operations release the carrier.

- StressStackOverflow is dialed down to run for 1m instead of 2mins.

- The use of CountDownLatch in a number of tests that poll thread state has been dropped to keep the tests as simple as possible.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8320707](https://bugs.openjdk.org/browse/JDK-8320707): Virtual thread test updates (**Enhancement** - P4)

### Reviewers
 * [Jaikiran Pai](https://openjdk.org/census#jpai) (@jaikiran - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17136/head:pull/17136` \
`$ git checkout pull/17136`

Update a local copy of the PR: \
`$ git checkout pull/17136` \
`$ git pull https://git.openjdk.org/jdk.git pull/17136/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17136`

View PR using the GUI difftool: \
`$ git pr show -t 17136`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17136.diff">https://git.openjdk.org/jdk/pull/17136.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17136#issuecomment-1859062598)

### CODE REVIEW

**Comment by plummercj** [comment]
Can you explain your motivation for using AtomicBoolean with a polling loop rather than CountDownLatch(1)? I'm working on a test where I just added a CountDownLatch(1) and am wondering if I should do the same, but I'm not sure if there is something about these tests that is motivating the change.

**Comment by AlanBateman** [comment]
> I'm working on a test where I just added a CountDownLatch(1) and am wondering if I should do the same, but I'm not sure if there is something about these tests that is motivating the change.

CountDownLatch is great for many tests. It's not as powerful as a Phaser of course but good enough and usually easy to understand quickly what is going on. However, for tests that are testing thread state then you often want to have as few dependencies as possible. In the case of CNL, await method may park. The countDown method may have to unpark waiters for for virtual threads it means potentially parking (to queue a task) when unparking, so temporary transitions that JVMTI has to be concerned with. The other thing is keeping tests simple/consistent, it can be hard to maintain tests where one test method coordinates with one approach, another test method does something different.

File: test/jdk/java/lang/Thread/virtual/stress/Skynet.java [hunk 1]
```diff
@@ -26,7 +26,7 @@
  * @summary Stress test virtual threads with a variation of the Skynet 1M benchmark
  * @requires vm.continuations
  * @requires !vm.debug | vm.gc != "Z"
- * @run main/othervm/timeout=300 -Xmx1g Skynet
+ * @run main/othervm/timeout=300 -Xmx1500m Skynet
```
Comment (jaikiran) [inline test/jdk/java/lang/Thread/virtual/stress/Skynet.java:29]: Are these heap sizing changes to reduce the resource usage of this test or is it to try and trigger any potential issue that this test verifies?

**Review by jaikiran — APPROVED** [review]
Overall these changes look OK to me. I haven't paid much attention to the jvmti test change, but I've looked at others and they seem fine.
The `StressStackOverflow` test has its duration reduced by half which appears intentional and then there's a new test library helper class to allow pinning tasks in tests.

Just one question - doesn't the use of a new native code in the test library (the `libVThreadPinner.c`) require any build changes (I'm not too familiar with that part)?

File: test/jdk/java/lang/Thread/virtual/stress/Skynet.java (hunk 1 above)
Comment (AlanBateman) [reply test/jdk/java/lang/Thread/virtual/stress/Skynet.java:29]: > Are these heap sizing changes to reduce the resource usage of this test or is it to try and trigger any potential issue that this test verifies?

The heap usage for this one can vary quite a bit, some of the lower optimization modes lead to more contention so there are are more FJP submissions. See JDK-8303635 for one example. It's been bumped to 1.5g in the loom repo for some time.

**Comment by AlanBateman** [comment]
> Just one question - doesn't the use of a new native code in the test library (the `libVThreadPinner.c`) require any build changes (I'm not too familiar with that part)?

Magnus did some work in the make files some time ago to build native libs in the test/lib tree.

File: test/jdk/java/lang/Thread/virtual/stress/Skynet.java (hunk 1 above)
Comment (jaikiran) [reply test/jdk/java/lang/Thread/virtual/stress/Skynet.java:29]: >  It's been bumped to 1.5g in the loom repo for some time.

Ah, I misread that change previously as reducing it from `1g` to `150m`, while infact this change is increasing it to `1500m`. Given your reference to JDK-8303635, the change looks OK to me.

**Pull request closed without merging**